aws/credentials/processcreds/provider.go

```diff
@@ -0,0 +1,476 @@
+
+	return credentials.NewCredentials(p)
+}
+
+// NewCredentialsTimeout returns a pointer to a new Credentials object with
+// the specified command and timeout, and default duration and max buffer size.
+func NewCredentialsTimeout(command string, timeout time.Duration) *credentials.Credentials {
+	p := NewCredentials(command, func(opt *ProcessProvider) {
+		opt.Timeout = timeout
+	})
+
+	return p
+}
+
+type credentialProcessResponse struct {
+	Version         int
+	AccessKeyID     string `json:"AccessKeyId"`
+	SecretAccessKey string
+	SessionToken    string
+	Expiration      *time.Time
+}
+
+// Retrieve executes the 'credential_process' and returns the credentials.
+func (p *ProcessProvider) Retrieve() (credentials.Value, error) {
+	out, err := p.executeCredentialProcess()
+	if err != nil {
+		return credentials.Value{ProviderName: ProviderName}, err
+	}
+
+	// Serialize and validate response
+	resp := &credentialProcessResponse{}
+	if err = json.Unmarshal(out, resp); err != nil {
+		return credentials.Value{ProviderName: ProviderName}, awserr.New(
+			ErrCodeProcessProviderParse,
+			fmt.Sprintf("%s: %s", ErrMsgProcessProviderParse, string(out)),
+			err)
+	}
+
+	if resp.Version != 1 {
+		return credentials.Value{ProviderName: ProviderName}, awserr.New(
+			ErrCodeProcessProviderVersion,
+			ErrMsgProcessProviderVersion,
+			nil)
+	}
+
+	if len(resp.AccessKeyID) == 0 {
+		return credentials.Value{ProviderName: ProviderName}, awserr.New(
+			ErrCodeProcessProviderRequired,
+			ErrMsgProcessProviderMissKey,
+			nil)
+	}
+
+	if len(resp.SecretAccessKey) == 0 {
+		return credentials.Value{ProviderName: ProviderName}, awserr.New(
+			ErrCodeProcessProviderRequired,
+			ErrMsgProcessProviderMissSecret,
+			nil)
+	}
+
+	// Handle expiration
+	p.staticCreds = resp.Expiration == nil
+	if resp.Expiration != nil {
+		p.SetExpiration(*resp.Expiration, p.ExpiryWindow)
+	}
+
+	return credentials.Value{
+		ProviderName:    ProviderName,
+		AccessKeyID:     resp.AccessKeyID,
+		SecretAccessKey: resp.SecretAccessKey,
+		SessionToken:    resp.SessionToken,
+	}, nil
+}
+
+// IsExpired returns true if the credentials retrieved are expired, or not yet
+// retrieved.
+func (p *ProcessProvider) IsExpired() bool {
+	if p.staticCreds {
+		return false
+	}
+	return p.Expiry.IsExpired()
+}
+
+func findInSlice(needle string, haystack []string) bool {
+	for _, straw := range haystack {
+		if strings.HasPrefix(straw, needle) {
+			return true
+		}
+	}
+	return false
+}
+
+func getProcessProviderEnv(existEnv []string) []string {
+	if !findInSlice(AlreadyPreparedCommandKey, existEnv) {
+		existEnv = append(existEnv, fmt.Sprintf(
+			"%s=%s",
+			AlreadyPreparedCommandKey,
+			"true"))
+
+		if path, ok := os.LookupEnv("PATH"); !ok || len(path) < 1 {
+			if runtime.GOOS == GoOSWindowsKey {
+				existEnv = append(existEnv, fmt.Sprintf(
+					"%s=%s",
+					"PATH",
+					"C:\\Windows\\system32"))
+			} else {
+				existEnv = append(existEnv, fmt.Sprintf("%s=%s", "PATH", "/bin"))
```

Comment:
Why do the system folders need to be added to the path? Wouldn't the command execute in the same environment as the parent?

Reply:
I only add the system32 dir to the path if it's not already in the path. 99.9% of the time, it should already be on the path. The main time it's not on the path is during testing when the env gets clobbered.